### PR TITLE
sql: new functions `crdb_internal.generate_test_objects` and `.gen_rand_ident`

### LINKED ISSUE #94026: sql: generate many tables quickly
We would like a way to generate many tables (and schemas, and dbs) very quickly.
This will facilitate testing and help the observability teams design scalable APIs.

Jira issue: CRDB-22647
Epic: CRDB-20388

### PR DESCRIPTION
**Pull request by knz**
Fixes #94026.

tldr: the new function
```
   crdb_internal.generate_test_objects(<parameters>)
```
generates multiple SQL objects at once. It is meant for use by tests.

Additionally, a new flag `--expand-schema` was added to `cockroach
demo` which exercises and demonstrates this new feature. When
specified, `demo` will gradually *multiply* the workload schema
by the specified number of schema objects, by adding databases
and tables similar to the one defined by the workload.

For example:

```
$ cockroach demo movr --expand-schema=10000
... wait a bit ...

> SHOW DATABASES;
... observe: similar databases to movr have been created, 100 tables each
```

----

The parameters argument is a JSONB object, containing at least the
following fields:

- `"counts"`: the counts of objects to create.
- `"names"`: the pattern to use to name generated objects.

The pattern and counts array influences object creation as follows:

| Counts     | Pattern     | Creates databases? | Creates schemas?          | Creates tables?                           | Total objects created                 |
|------------|-------------|--------------------|---------------------------|-------------------------------------------|---------------------------------------|
| N          | `tb`        | no                 | no                        | yes: `<currentdb>.<currentschema>.tb1`... | N                                     |
| N          | `sc.tb`     | no                 | no                        | yes: `<currentdb>.sc.tb1`...              | N                                     |
| N          | `db.sc.tb`  | no                 | no                        | yes: `db.sc.tb1`...                       | N                                     |
| N2, N1     | `sc.tb`     | no                 | yes: `<currentdb>.sc1`... | yes: `<currentdb>.sc1.tb1`...             | N2 + (N1 x N2)                        |
| N2, 0      | `sc.tb`     | no                 | yes: `<currentdb>.sc1`... | no                                        | N2                                    |
| N2, N1     | `db.sc.tb`  | no                 | yes: `db.sc1`...          | yes: `db.sc1.tb1`...                      | N2 + (N1 x N2)                        |
| N3, N2, N1 | `db.sc.tb`  | yes: `db1`...      | yes: `db1.sc1`...         | yes: `db1.sc1.tb1`...                     | (N3 x 2) + (N3 x N2) + (N1 x N2 x N3) |
| N3, N2, 0  | `db.sc._`   | yes: `db1`...      | yes: `db1.sc1`...         | no                                        | (N3 x 2) + (N3 x N2)                  |
| N2, 0      | `db.sc._`   | no                 | yes: `db.sc1`...          | no                                        | N2                                    |
| N3, 0, 0   | `db._._`    | yes: `db1`...      | no                        | no                                        | N3 x 2                                |
| N1, 0, N2  | `db.tb`     | yes: `db1`...      | no                        | yes: `db1.public.tb1`...                  | (N3 x 2) + (N3 x N2)                  |
| 0          | `tb`        | no                 | no                        | no                                        | 0                                     |
| 0, N2      | `prefix.tb` | no                 | no                        | no                                        | 0                                     |
| 0, N2, N1  | `db.sc.tb`  | no                 | no                        | no                                        | 0                                     |
| 0          | `prefix.tb` | no                 | no                        | no                                        | 0                                     |
| 0          | `db.sc.tb`  | no                 | no                        | no                                        | 0                                     |
| 0, N1      | `db.sc._`   | no                 | no                        | no                                        | 0                                     |

NB: the reason why the requested number of databases is doubled in the
total count of objects created is that a `public` schema is created for
each new database.

The convenience overload `generate_test_objects(pattern : string,
count : integer)` is also provided as a convenience alias for
`generate_test_objects('{"names":pattern, "counts":[count]}'::JSONB)`.

Likewise, `generate_test_objects(pattern : string, counts :
[]integer)` is also provided as a convenience alias for
`generate_test_objects('{"names":pattern, "counts":counts}'::JSONB)`.

The configuration parameters may also contain the following fields:

- `"names"`: pattern to use to name the generated objects (default:
  "_", see note about `"table_templates"` below).
- `"counts"`: counts of generated objects (default: [10]).
- `"dry_run"`: prepare the schema but do not actually write it
  (default: false).
- `"seed"`: random seed to use (default: auto).
- `"randomize_columns"`: whether to randomize the column names on tables
  (default: true).
- `"table_templates"`: table templates to use.
  If the last part of `"names"` is "_", the name of the template
  will be used as base pattern during name generation for tables.
  Otherwise, the last part of "names" will be used as pattern.
  If no table templates are specified, a simple template is used.
- `"name_gen"`: configuration for the name generation, see below.

Guaranteed properties:

- when a requested object count is zero, no object of that type (and
  no object of sub-types) gets created. This preserves the
  mathematical properties:
  - calls with a count of zero are idempotent.
  - if the function is called sequentially
    with counts N and M, the total number of objects is always N+M
    regardless of the values of N and M.

Also for testing and troubleshooting, another function is provided:
```
crdb_internal.gen_rand_ident(name_pattern: string, count: int, parameters: jsonb) -> string
```
It produces SQL identifiers using the same algorithm as used by
`generate_test_objects`.

Its parameters are:

- `"seed"`: the seed to use for the pseudo-random generator (default:
  random).
-` "number"`: whether to add a number to the generated names (default
  true). When enabled, occurrences of the character '#' in the name
  pattern are replaced by the number. If '#' is not present, the
  number is added at the end.
- `"noise"`: whether to add noise to the generated names (default
  true). It adds a non-zero probability for each of the probability
  options below left to zero. (To enable noise generally but disable
  one type of noise, set its probability to -1.)
- `"punctuate"`: probability of adding punctuation to the generated
  names.
- `"quote"`: probabiltiy of adding single or double quotes to the
  generated names.
- `"emote"`: probability of adding emojis to the generated names.
- `"space"`: probability of adding simple spaces to the generated
  names.
- `"whitespace"`: probability of adding complex whitespace to the
  generated names.
- `"capitals"`: probability of using capital letters in the generated
  names. Note: the name pattern must contain ASCII letters already for
  capital letters to be used.
- `"diacritics"`: probability of adding diacritics in the generated
  names.
- `"diacritic_depth"`: max number of diacritics to add at a time
  (default 1).
- `"zalgo"`: special option that overrides diacritics and
  diacritic_depth (default false).

For convenience, the overload `crdb_internal.gen_rand_ident(name_pattern: string, count: int)`
is provided as an alias to `gen_rand_ident(pattern, count, '{}'::JSONB)`.

Release note: None

### CODE REVIEW

**Comment by dt** [comment]
two quick drive-by ergonomics questions:
* How would you feel about including "test" or "testing" in the name, to make its purpose clearer, e.g. "crdb_internal.generate_test_objects" ?
* How would you feel about different arity overloads instead of an array of variable length, so that the parameters can be named?

**Comment by knz** [comment]
>   How would you feel about including "test" or "testing" in the name, to make its purpose clearer, e.g. "crdb_internal.generate_test_objects" ?

no objection

>   How would you feel about different arity overloads instead of an array of variable length, so that the parameters can be named?

let's delay this convo until I show you the degree of configurability i'm aiming for.

**Comment by knz** [comment]
@dt okay can you look again at the proposed interface (see updated PR description)

Now I am willing to discuss UX. I'd also like the common case to be simple to type.

**Comment by knz** [comment]
So this is good for a first round of review, tests and all.

@postamar @ajwerner I'm thinking the schema team would like to own this; how do you feel about that?

File: pkg/sql/catalog/randgen/api.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+// TestSchemaGeneratorConfig is an alias for randgencfg.Config.
+type TestSchemaGeneratorConfig = randgencfg.Config
+
+// TestSchemaGenerator can be used to generate random SQL objects.
+// See NewTestSchemaGenerator().
+//
+// A design goal for this facility is to ensure that the generation is
+// fast in the common case, so it can be used to generate large
+// schemas quickly.
+type TestSchemaGenerator interface {
```
Comment (postamar) [inline pkg/sql/catalog/randgen/api.go:30]: Should this be named `RandomSchemaGenerator` instead? If so, propagate the name change.

File: pkg/sql/catalog/randgen/randgen.go [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+
+	// b is the current batch.
+	b *kv.Batch
+	// newDescs are the recently added descriptors that need a new
+	// namespace entry.
+	newDescs []catalog.MutableDescriptor
+	// queuedDescs are the recently updated descriptors that need to be
+	// written to KV.
+	queuedDescs []catalog.MutableDescriptor
+}
```
Comment (postamar) [inline pkg/sql/catalog/randgen/randgen.go:225]: I would find it useful to group the fields in this struct according to their lifecycle/mutability, either by rearranging the lines in which they appear, or by introducing sub-structs. Some fields are part of the "generator state" so to speak (like `baseSchema`, `rand`), some are part of the generator output (`b` etc), some are immutable configuration values (`cfg`, `kvtrace` etc) and some are dependencies (`cat`, `coll`, etc).

File: pkg/sql/catalog/randgen/tables.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+			}
+		}
+	}
+	return selectTmpl, selectNames
+}
+
+func (g *testSchemaGenerator) genOneTable(
+	ctx context.Context,
+	parentDb catalog.DatabaseDescriptor,
+	parentSc catalog.SchemaDescriptor,
```
Comment (postamar) [inline pkg/sql/catalog/randgen/tables.go:152]: uber nit: FYI the prevailing convention in `pkg/sql/catalog/...` is to just call these `db` and `sc`

File: pkg/sql/catalog/randgen/templates.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+	// coming from the go map order.
+	sobjIDs := make([]descpb.ID, 0, len(objIDs))
+	for id := range objIDs {
+		sobjIDs = append(sobjIDs, id)
+	}
+	sort.Slice(sobjIDs, func(i, j int) bool { return sobjIDs[i] < sobjIDs[j] })
+
+	// Look up the descriptors from the IDs.
+	descs, err := g.coll.GetImmutableDescriptorsByID(ctx, g.txn,
+		tree.CommonLookupFlags{Required: true},
```
Comment (postamar) [inline pkg/sql/catalog/randgen/templates.go:87]: nit: `Required: true` is always implied by `GetImmutableDescriptorsByID`. One of the many sad little warts of the collection API which I hope to get rid of soon, see https://github.com/cockroachdb/cockroach/pull/93813 if you're interested in this.

File: pkg/sql/catalog/randgen/templates.go [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+				ID:              0,
+				Name:            "primary",
+				ColumnNames:     []string{"rowid"},
+				ColumnIDs:       []descpb.ColumnID{1},
+				DefaultColumnID: 1,
+			},
+		},
+		PrimaryIndex: descpb.IndexDescriptor{
+			ID:                  1,
+			Name:                "pk",
```
Comment (postamar) [inline pkg/sql/catalog/randgen/templates.go:205]: Consider using `tabledesc.PrimaryKeyIndexName` here

File: pkg/sql/catalog/randgen/util.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+// scNamesInDb retrieves the list of schema names in the given
+// database.
+func (g *testSchemaGenerator) scNamesInDb(
+	ctx context.Context, db catalog.DatabaseDescriptor,
+) map[string]struct{} {
+	res := make(map[string]struct{})
+	_ = db.ForEachSchema(func(_ descpb.ID, name string) error {
+		res[name] = struct{}{}
+		return nil
+	})
```
Comment (postamar) [inline pkg/sql/catalog/randgen/util.go:57]: Consider using `coll.GetAllSchemasInDatabase` here, unless you don't care that this won't work when temporary schemas are involved.

File: pkg/sql/catalog/randgen/tables.go [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+			colName := ng.GenerateOne(0)
+			tmpl.desc.Columns[i+1].Name = colName
+			tmpl.desc.Families[0].ColumnNames[i+1] = colName
+		}
+		ng := randident.NewNameGenerator(&nameGenCfg, g.rand, "primary")
+		idxName := ng.GenerateOne(0)
+		tmpl.desc.PrimaryIndex.Name = idxName
+	}
+
+	tb := tabledesc.NewBuilder(&tmpl.desc).BuildCreatedMutableTable()
```
Comment (postamar) [inline pkg/sql/catalog/randgen/tables.go:193]: You should call `RunPostDeserializationChanges` on the builder before building the descriptor here and elsewhere to make sure this functionality ages gracefully.

File: pkg/sql/catalog/randgen/batch.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+func (g *testSchemaGenerator) newDesc(ctx context.Context, desc catalog.MutableDescriptor) {
+	if g.cfg.DryRun {
+		return
+	}
+	g.newDescs = append(g.newDescs, desc)
+	g.queueDescMut(ctx, desc)
+}
+
+// queueDesc queues a modified descriptor for (later) write to KV.
+func (g *testSchemaGenerator) queueDescMut(ctx context.Context, desc catalog.MutableDescriptor) {
```
Comment (postamar) [inline pkg/sql/catalog/randgen/batch.go:30]: I think it's worth calling `validate.Self(clusterversion.TestingClusterVersion, desc)` right here. Validation will be performed anyway when the transaction commits but since a lot of descriptors are involved, _if_ a validation failure occurs it might be difficult to make sense of it.

**Review by postamar — APPROVED** [review]
Thanks for doing this. This LGTM modulo a couple of easy fixes related to building and validating descriptors. The rest is just nits, questions and suggestions, but nothing I wouldn't be OK with being merged.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94027)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[`pkg/sql/catalog/randgen/api.go` line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94027#-NK5UXsvr67i2hLkqHVd-r2-30:-NL0MIL_5bC5Dhqk7Sut:b-6m526x) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d092b6a47c6c12e20bf1e75f94b057e73daf30/pkg/sql/catalog/randgen/api.go#L30)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Should this be named `RandomSchemaGenerator` instead? If so, propagate the name change.
</blockquote></details>

While the names are randomized, the structure of the schema is not. We should _also_ have such a thing as a _random schema_ generator but this is not it.
___
*[`pkg/sql/catalog/randgen/batch.go` line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94027#-NK5_ip2ysG8CU6OUREI-r2-30:-NL0xnuV3oD_wpIyHiyC:b-77ep0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d092b6a47c6c12e20bf1e75f94b057e73daf30/pkg/sql/catalog/randgen/batch.go#L30)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I think it's worth calling `validate.Self(clusterversion.TestingClusterVersion, desc)` right here. Validation will be performed anyway when the transaction commits but since a lot of descriptors are involved, _if_ a validation failure occurs it might be difficult to make sense of it.
</blockquote></details>

I don't buy it. If validation failure occurs, it's only because of some property of the templates and that would show up at just 1 descriptor created.
___
*[`pkg/sql/catalog/randgen/randgen.go` line 117 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94027#-NK5UXsvr67i2hLkqHVe-r2-117:-NL0PoQV-BMVRGDeyqnh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d092b6a47c6c12e20bf1e75f94b057e73daf30/pkg/sql/catalog/randgen/randgen.go#L117)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Should this error come with a hint?
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/randgen/randgen.go` line 123 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94027#-NK5UXsvr67i2hLkqHVe-r2-123:-NL0PpPcCJcHMGLkoC24:b-kfkdqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d092b6a47c6c12e20bf1e75f94b057e73daf30/pkg/sql/catalog/randgen/randgen.go#L123)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Dumb question perhaps but is there any point in allowing non-admin users to bulk generate tests objects in the first place?
</blockquote></details>

You tell me :)
The view i've taken so far is that this code is not doing anything that a user couldn't do via CREATE statements in the first place. Only the _rate_ is different, although again a client could achieve similar things via concurrent client connections.
Maybe we could ask a PM for advice?
___
*[`pkg/sql/catalog/randgen/randgen.go` line 160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94027#-NK5UXsvr67i2hLkqHVe-r2-160:-NL0QijKCH0zer_0cRmv:b-47pj0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d092b6a47c6c12e20bf1e75f94b057e73daf30/pkg/sql/catalog/randgen/randgen.go#L160)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Can we make this interface package-private, if possible? There's no benefit to exporting it, is there?

My concern here is that exporting something yet again named "catalog" potentially makes the codebase more confusing.
</blockquote></details>

I don't understand how to do this. Doesn't the linter require that all types in a public function/method are also public?
___
*[`pkg/sql/catalog/randgen/randgen.go` line 225 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94027#-NK5UXsvr67i2hLkqHVe-r2-225:-NL0QqumECzsTv5R3BJI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d092b6a47c6c12e20bf1e75f94b057e73daf30/pkg/sql/catalog/randgen/randgen.go#L225)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I would find it useful to group the fields in this struct according to their lifecycle/mutability, either by rearranging the lines in which they appear, or by introducing sub-structs. Some fields are part of the "generator state" so to speak (like `baseSchema`, `rand`), some are part of the generator output (`b` etc), some are immutable configuration values (`cfg`, `kvtrace` etc) and some are dependencies (`cat`, `coll`, etc).
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/randgen/tables.go` line 152 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94027#-NK5_ip38clYb3692fIR-r2-152:-NL0XXWt05P1cVDKOfSQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d092b6a47c6c12e20bf1e75f94b057e73daf30/pkg/sql/catalog/randgen/tables.go#L152)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

uber nit: FYI the prevailing convention in `pkg/sql/catalog/...` is to just call these `db` and `sc`
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/randgen/tables.go` line 193 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94027#-NK5_ip38clYb3692fIR-r2-193:-NL0Z1Jz3pCkvKFpSu7k:b-ji7fqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d092b6a47c6c12e20bf1e75f94b057e73daf30/pkg/sql/catalog/randgen/tables.go#L193)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

You should call `RunPostDeserializationChanges` on the builder before building the descriptor here and elsewhere to make sure this functionality ages gracefully.
</blockquote></details>

Absolutely not. This causes the performance to tank by 50% (those checks are _expensive_!!!) and for no good reason:
1) it's the same template used every time. 
2) we're 100% controlling the template here. If there's a bug, it would be caught by the test suite.
___
*[`pkg/sql/catalog/randgen/templates.go` line 205 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94027#-NKAIOK7Xd-J1rxN5cYh-r2-205:-NL0_ENb4SW5rHtev-qJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d092b6a47c6c12e20bf1e75f94b057e73daf30/pkg/sql/catalog/randgen/templates.go#L205)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Consider using `tabledesc.PrimaryKeyIndexName` here
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/randgen/util.go` line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94027#-NK5_ip38clYb3692fIS-r2-57:-NL0_FqI8uTiGZoJdROJ:bwh7osx) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d092b6a47c6c12e20bf1e75f94b057e73daf30/pkg/sql/catalog/randgen/util.go#L57)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Consider using `coll.GetAllSchemasInDatabase` here, unless you don't care that this won't work when temporary schemas are involved.
</blockquote></details>

I do care about temporary schemas and this is exercised in tests.
___
*[`pkg/sql/sem/builtins/builtins.go` line 6434 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94027#-NJmFtwqgirzyntn4Mqj-r2-6434:-NL0b6bz07GoL5E9Hqhz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d092b6a47c6c12e20bf1e75f94b057e73daf30/pkg/sql/sem/builtins/builtins.go#L6434)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Would it be possible to redefine this and the overload above as UDFs instead?
</blockquote></details>

Done.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 575 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94027#-NK4xGT5KjZwrcXpZUWz-r2-575:-NL0b7Nf6SyaMzOm-gOC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d092b6a47c6c12e20bf1e75f94b057e73daf30/pkg/sql/sem/builtins/generator_builtins.go#L575)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Same comment as above: can this overload be defined as a UDF instead?
</blockquote></details>

Done.
___
*[`pkg/sql/sem/eval/deps.go` line 220 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94027#-NJmFtwqgirzyntn4Mql-r2-220:-NL0xfHg8aROuuvnGL09:bw6j4gg) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d092b6a47c6c12e20bf1e75f94b057e73daf30/pkg/sql/sem/eval/deps.go#L220)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I get why the param and the return value are strings but I wonder if `interface{}` wouldn't be more convenient. What do you think? I don't feel strongly about this, in any case.
</blockquote></details>

I've tried it the other ways, and it results in code that's slightly harder to read/maintain.
___
*[`pkg/sql/logictest/testdata/logic_test/rand_ident` line 12 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94027#-NK4xGT5KjZwrcXpZUWr-r2-12:-NL0z2jUAizT9vteRqr6:btukf5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d092b6a47c6c12e20bf1e75f94b057e73daf30/pkg/sql/logictest/testdata/logic_test/rand_ident#L12)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: Could `seed` be `omitempty` and its value be 0 in tests? It's no worse than 123 which is used everywhere (I think)
</blockquote></details>

I've made it omitempty because zero is, indeed, a valid rng seed value and we shouldn't treat it as special.
However, I do see value in using "123" in tests, it has better explanatory power (using "0" would make the user wonder if it's special in some way").
___
*[`pkg/sql/logictest/testdata/logic_test/test_objects` line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94027#-NK7u1GHKonWVqKzRpLE-r2-1:-NL0_oAQ5VtYLYs28jx8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d092b6a47c6c12e20bf1e75f94b057e73daf30/pkg/sql/logictest/testdata/logic_test/test_objects#L1)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Consider renaming this file to `gen_test_objects` perhaps? `test_objects` can give the impression that this is about testing objects.
</blockquote></details>

Done.
___
*[`pkg/util/randident/namegen_test.go` line 110 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94027#-NK53a0WAeXAuEZ-qIjU-r2-110:-NL0LxZZ42M5icZmI8Zo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d092b6a47c6c12e20bf1e75f94b057e73daf30/pkg/util/randident/namegen_test.go#L110)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: perhaps having `testNameGen` take a `struct{always, never, sometimes []pred}` type might be simpler?
</blockquote></details>

Done.

**Comment by knz** [comment]
i've made it admin-only by default like discussed.

**Comment by knz** [comment]
bors r=postamar

**Comment by postamar** [comment]
Re. `RunPostDeserializationChanges`: instead of running it on all the descriptors you generate, you could run it only on the template, before using that proto for instantiation. It's easier said than done so I'll follow up with a PR.

I'm fine with all your other decisions.

**Comment by postamar** [comment]
See https://github.com/cockroachdb/cockroach/pull/94840 to that effect. As it turns out, the test actually picked up something. Nothing worth worrying about, but enough to justify its existence.
